### ISSUE TITLE
Migrations :: exception thrown when changing key from regular to composite and adding navigation property at the same time

### ISSUE THREAD

**Issue by maumar** [dotnet/efcore#4804]
Don't know exactly what causes the issue. Hopefully this will be enough to reproduce it later:

Previous migration designer file:

```
    [DbContext(typeof(PoeSniperContext))]
    [Migration("20160315184024_Mig02")]
    partial class Mig02
    {
        protected override void BuildTargetModel(ModelBuilder modelBuilder)
        {
            modelBuilder
                .HasAnnotation("ProductVersion", "1.0.0-rc2-20207")
                .HasAnnotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn);

            modelBuilder.Entity("Model.Account", b =>
                {
                    b.Property<string>("AccountName");

                    b.Property<string>("LastCharacterName");

                    b.HasKey("AccountName");

                    b.ToTable("Account");
                });

            modelBuilder.Entity("Model.Item", b =>
                {
                    b.Property<string>("Id");

                    b.Property<int>("Ilvl");

                    b.Property<string>("Name");

                    b.Property<string>("StashAccountName");

                    b.Property<string>("StashId");

                    b.Property<string>("TypeLine");

                    b.HasKey("Id");

                    b.HasIndex("StashId", "StashAccountName");

                    b.ToTable("Items");
                });

            modelBuilder.Entity("Model.ItemFeedChunk", b =>
                {
                    b.Property<string>("ChunkId");

                    b.Property<int>("Index");

                    b.Property<string>("PreviousChunkId");

                    b.Property<int?>("PreviousIndex");

                    b.HasKey("ChunkId", "Index");

                    b.HasIndex("PreviousChunkId", "PreviousIndex");

                    b.ToTable("ItemFeedChunks");
                });

            modelBuilder.Entity("Model.Stash", b =>
                {
                    b.Property<string>("Id");

                    b.Property<string>("AccountName");

                    b.Property<string>("ItemFeedChunkChunkId");

                    b.Property<int?>("ItemFeedChunkIndex");

                    b.HasKey("Id", "AccountName");

                    b.HasIndex("AccountName");

                    b.HasIndex("ItemFeedChunkChunkId", "ItemFeedChunkIndex");

                    b.ToTable("Stashes");
                });

            modelBuilder.Entity("Model.Item", b =>
                {
                    b.HasOne("Model.Stash")
                        .WithMany()
                        .HasForeignKey("StashId", "StashAccountName");
                });

            modelBuilder.Entity("Model.ItemFeedChunk", b =>
                {
                    b.HasOne("Model.ItemFeedChunk")
                        .WithOne()
                        .HasForeignKey("Model.ItemFeedChunk", "PreviousChunkId", "PreviousIndex");
                });

            modelBuilder.Entity("Model.Stash", b =>
                {
                    b.HasOne("Model.Account")
                        .WithMany()
                        .HasForeignKey("AccountName")
                        .OnDelete(DeleteBehavior.Cascade);

                    b.HasOne("Model.ItemFeedChunk")
                        .WithMany()
                        .HasForeignKey("ItemFeedChunkChunkId", "ItemFeedChunkIndex");
                });
        }
    }

```

Current model:

```
    public class Account
    {
        public string AccountName { get; set; }

        public List<Stash> Stashes { get; set; }

        public string LastCharacterName { get; set; }
    }

    public class Item
    {
        public string Id { get; set; }

        public int Ilvl { get; set; }

        public string Name { get; set; }

        public string TypeLine { get; set; }

        public bool Corrupted { get; set; }

        public List<ItemMod> ImplicitMods { get; set; }
    }

    public class ItemFeedChunk
    {
        public string ChunkId { get; set; }
        public int Index { get; set; }

        public string PreviousChunkId { get; set; }
        public int? PreviousIndex { get; set; }
        public ItemFeedChunk PreviousChunk { get; set; }

        public List<Stash> Stashes { get; set; }

    }

    public class ItemMod
    {
        public int Id { get; set; }
        public ItemModName Name { get; set; }
        public decimal Value { get; set; }
    }

    public class ItemModName
    {
        public int Id { get; set; }
        public string Text { get; set; }
    }

    public class MagicalItem : Item
    {
        //public List<ItemMod> ExplicitMods { get; set; }

        public bool Identified { get; set; }
    }

    public class PoeSniperContext : DbContext
    {
        public DbSet<Stash> Stashes { get; set; }
        public DbSet<ItemFeedChunk> ItemFeedChunks { get; set; }
        public DbSet<Item> Items { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            var connectionString = new SqlConnectionStringBuilder
            {
                DataSource = ".",
                InitialCatalog = "PoeSniperTest",
                MultipleActiveResultSets = true,
                IntegratedSecurity = true
            }.ToString();

            optionsBuilder.UseSqlServer(connectionString);
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<ItemFeedChunk>().HasKey(e => new { e.ChunkId, e.Index });
            modelBuilder.Entity<ItemFeedChunk>().HasOne(e => e.PreviousChunk).WithOne().HasForeignKey<ItemFeedChunk>(e => new { e.PreviousChunkId, e.PreviousIndex }).IsRequired(false);

            modelBuilder.Entity<Account>().HasKey(e => e.AccountName);

            modelBuilder.Entity<Stash>().HasKey(e => new { e.Id, e.AccountName });
            modelBuilder.Entity<Stash>().HasOne(e => e.Account).WithMany(e => e.Stashes).HasForeignKey(e => e.AccountName);

            modelBuilder.Entity<Item>().HasKey(e => new { e.Id, e.Name });
            modelBuilder.Entity<MagicalItem>().HasBaseType<Item>();
            modelBuilder.Entity<UniqueItem>().HasBaseType<MagicalItem>();
        }
    }

    public class Stash
    {
        public string Id { get; set; }
        public string AccountName { get; set; }

        public Account Account { get; set; }
        public List<Item> Items { get; set; }
    }

    public class UniqueItem : MagicalItem
    {
        public string FlavorText { get; set; }
    }
```

Failing migration code:

```
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.DropPrimaryKey(
                name: "PK_Items",
                table: "Items");

            migrationBuilder.CreateTable(
                name: "ItemModName",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
                    Text = table.Column<string>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_ItemModName", x => x.Id);
                });

            migrationBuilder.CreateTable(
                name: "ItemMod",
                columns: table => new
                {
                    Id = table.Column<int>(nullable: false)
                        .Annotation("SqlServer:ValueGenerationStrategy", SqlServerValueGenerationStrategy.IdentityColumn),
                    ItemId = table.Column<string>(nullable: true),
                    ItemName = table.Column<string>(nullable: true),
                    NameId = table.Column<int>(nullable: true),
                    Value = table.Column<decimal>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("PK_ItemMod", x => x.Id);
                    table.ForeignKey(
                        name: "FK_ItemMod_ItemModName_NameId",
                        column: x => x.NameId,
                        principalTable: "ItemModName",
                        principalColumn: "Id",
                        onDelete: ReferentialAction.Restrict);
                    table.ForeignKey(
                        name: "FK_ItemMod_Items_ItemId_ItemName",
                        columns: x => new { x.ItemId, x.ItemName },
                        principalTable: "Items",
                        principalColumns: new[] { "Id", "Name" },
                        onDelete: ReferentialAction.Restrict);
                });

            migrationBuilder.AlterColumn<string>(
                name: "Name",
                table: "Items",
                nullable: false);

            migrationBuilder.AddColumn<bool>(
                name: "Corrupted",
                table: "Items",
                nullable: false,
                defaultValue: false);

            migrationBuilder.AddColumn<string>(
                name: "Discriminator",
                table: "Items",
                nullable: false,
                defaultValue: "");

            migrationBuilder.AddColumn<bool>(
                name: "Identified",
                table: "Items",
                nullable: true);

            migrationBuilder.AddColumn<string>(
                name: "FlavorText",
                table: "Items",
                nullable: true);

            migrationBuilder.AddPrimaryKey(
                name: "PK_Items",
                table: "Items",
                columns: new[] { "Id", "Name" });

            migrationBuilder.CreateIndex(
                name: "IX_ItemMod_NameId",
                table: "ItemMod",
                column: "NameId");

            migrationBuilder.CreateIndex(
                name: "IX_ItemMod_ItemId_ItemName",
                table: "ItemMod",
                columns: new[] { "ItemId", "ItemName" });
        }
```

**Comment by maumar** [comment]
Error thrown:
System.Data.SqlClient.SqlException (0x80131904): There are no primary or candidate keys in the referenced table 'Items' that match the referencing column list in the foreign key 'FK_ItemMod_Items_ItemId_ItemName'.
Could not create constraint. See previous errors.
   at System.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
   at System.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
   at System.Data.SqlClient.TdsParser.TryRun(RunBehavior runBehavior, SqlCommand cmdHandler, SqlDataReader dataStream, BulkCopySimpleResultSet bulkCopyHandler, TdsParserStateObject stateObj, Boolean& dataReady)
   at System.Data.SqlClient.SqlCommand.RunExecuteNonQueryTds(String methodName, Boolean async, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.InternalExecuteNonQuery(TaskCompletionSource`1 completion, String methodName, Boolean sendToPipe, Int32 timeout, Boolean asyncWrite)
   at System.Data.SqlClient.SqlCommand.ExecuteNonQuery()
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.Execute(IRelationalConnection connection, String executeMethod, IReadOnlyDictionary`2 parameterValues, Boolean openConnection, Boolean closeConnection)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues, Boolean manageConnection)
   at Microsoft.EntityFrameworkCore.Storage.Internal.RelationalCommandExtensions.ExecuteNonQuery(IEnumerable`1 commands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Execute(IEnumerable`1 relationalCommands)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.MigrationsOperations.UpdateDatabase(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_1.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)

**Comment by smitpatel** [comment]
Sorting issue! We are defining FK in create table call & the principal key used by fk is being defined later (as PK)

**Comment by bricelam** [comment]
This may have been fixed by #5480

**Comment by rowanmiller** [comment]
This doesn't seem like a scenario folks would really hit. May already be fixed.
